File: bench/bench.cpp
// Latency and throughput benchmark for the GPU prover.
#include "../include/vkzg.h"
#include "bench_common.h"

#include <algorithm>
#include <cstdio>
#include <cstring>
#include <vector>

int main() {
    const int reps = 10;
    const uint32_t maxBatch = 256;

    // Set to the highest blob count in the sweep, so every measurement below
    // exercises a single, real dispatch rather than being silently chunked.
    vkzg_options opts;
    vkzg_options_default(&opts);
    opts.max_batch_size = maxBatch;

    vkzg_prover *p = nullptr;
    vkzg_result rc = vkzg_prover_new(&p, &opts);
    if (rc != VKZG_OK) {
        printf("prover_new failed: %s\n", vkzg_error_string(rc));
        return 1;
    }
    printf("device: %s (%u GPU cores)\n\n", vkzg_prover_device_name(p),
           vkzg_prover_gpu_core_count(p));

    printf("Compute cell KZG proofs for N blobs:\n\n");

    const size_t proofBytes = (size_t)VKZG_NUM_CELL_PROOFS * VKZG_BYTES_PER_PROOF;

    std::vector<uint8_t> blobs((size_t)maxBatch * VKZG_BYTES_PER_BLOB);
    for (uint32_t i = 0; i < maxBatch; i++) {
        fill_blob(&blobs[(size_t)i * VKZG_BYTES_PER_BLOB], i + 1);
    }
    std::vector<uint8_t> proofs((size_t)maxBatch * proofBytes);

    auto run = [&](uint32_t n) {
        // warm up
        vkzg_compute_proofs_batch(p, proofs.data(), blobs.data(), n);
        double best = 1e30, total = 0;
        for (int r = 0; r < reps; r++) {
            double a = now_ms();
            vkzg_result e = vkzg_compute_proofs_batch(p, proofs.data(), blobs.data(), n);
            double ms = now_ms() - a;
            if (e != VKZG_OK) {
                printf("  blobs: %3u   error: %s\n", n, vkzg_error_string(e));
                return;
            }
            best = std::min(best, ms);
            total += ms;
        }
        printf("  blobs: %3u   best: %9.2f ms   avg: %9.2f ms   per blob: %9.2f ms\n", n, best,
               total / reps, best / n);
    };

    for (uint32_t n = 1; n <= maxBatch; n *= 2) {
        run(n);
    }

    // -------------------------------------------------- given 50% of cells for N blobs
    printf("\nRecover cells for N blobs given 50%% of cells:\n\n");

    const size_t cellsBytes = (size_t)VKZG_NUM_CELL_PROOFS * VKZG_BYTES_PER_CELL;
    std::vector<uint8_t> cells((size_t)maxBatch * cellsBytes);
    std::vector<uint8_t> present((size_t)maxBatch * VKZG_NUM_CELL_PROOFS);
    for (uint32_t i = 0; i < maxBatch; i++) {
        fill_cells(&cells[(size_t)i * cellsBytes], i + 1);
        // First 64 of 128 cells present, the rest missing -- a genuine
        // 50%-missing scenario (not the easier "every other cell" case).
        uint8_t *pr = &present[(size_t)i * VKZG_NUM_CELL_PROOFS];
        memset(pr, 1, VKZG_FIELD_ELEMENTS_PER_CELL);
        memset(pr + VKZG_FIELD_ELEMENTS_PER_CELL, 0, VKZG_NUM_CELL_PROOFS - VKZG_FIELD_ELEMENTS_PER_CELL);
    }
    std::vector<uint8_t> recovered((size_t)maxBatch * cellsBytes);

    auto runRecover = [&](uint32_t n) {
        vkzg_recover_cells_batch(p, recovered.data(), cells.data(), present.data(), n); // warm up
        double best = 1e30, total = 0;
        for (int r = 0; r < reps; r++) {
            double a = now_ms();
            vkzg_result e = vkzg_recover_cells_batch(p, recovered.data(), cells.data(), present.data(), n);
            double ms = now_ms() - a;
            if (e != VKZG_OK) {
                printf("  blobs: %3u   error: %s\n", n, vkzg_error_string(e));
                return;
            }
            best = std::min(best, ms);
            total += ms;
        }
        printf("  blobs: %3u   best: %9.2f ms   avg: %9.2f ms   per blob: %9.2f ms\n", n, best,
               total / reps, best / n);
    };

    for (uint32_t n = 1; n <= maxBatch; n *= 2) {
        runRecover(n);
    }

    vkzg_prover_free(p);
    return 0;
}

File: bindings/java/src/test/java/vkzg/Bench.java
package vkzg;

// Latency and throughput benchmark for the GPU prover -- the Java analog of
// bench/bench.cpp. Run with:
//   mvn test-compile
//   java --enable-native-access=ALL-UNNAMED -cp target/classes:target/test-classes vkzg.Bench
public class Bench {
    private static final int REPS = 10;
    private static final int MAX_BATCH = 256;

    public static void main(String[] args) {
        // Set to the highest blob count in the sweep, so every measurement below
        // exercises a single, real dispatch rather than being silently chunked.
        Vkzg.init(MAX_BATCH);
        System.out.printf("device: %s (%d GPU cores)%n%n", Vkzg.deviceName(), Vkzg.gpuCoreCount());

        System.out.println("Compute all cell KZG proofs for N blobs:\n");
        byte[] blobs = new byte[MAX_BATCH * Vkzg.BYTES_PER_BLOB];
        for (int i = 0; i < MAX_BATCH; i++) {
            fillBlob(blobs, i * Vkzg.BYTES_PER_BLOB, i + 1);
        }
        for (int n = 1; n <= MAX_BATCH; n *= 2) {
            runComputeProofs(blobs, n);
        }

        System.out.println("\nGiven 50% of cells for N blobs, recover the other half:\n");
        final int cellsPerBlobBytes = Vkzg.CELL_PROOFS_PER_BLOB * Vkzg.BYTES_PER_CELL;
        byte[] cells = new byte[MAX_BATCH * cellsPerBlobBytes];
        byte[] present = new byte[MAX_BATCH * Vkzg.CELL_PROOFS_PER_BLOB];
        for (int i = 0; i < MAX_BATCH; i++) {
            fillCells(cells, i * cellsPerBlobBytes, i + 1);
            // First 64 of 128 cells present, the rest missing -- a genuine
            // 50%-missing scenario (not the easier "every other cell" case).
            int base = i * Vkzg.CELL_PROOFS_PER_BLOB;
            for (int c = 0; c < 64; c++) {
                present[base + c] = 1;
            }
        }
        for (int n = 1; n <= MAX_BATCH; n *= 2) {
            runRecoverCells(cells, present, n);
        }

        Vkzg.deinit();
    }

    private static void runComputeProofs(byte[] blobs, int n) {
        byte[] slice = java.util.Arrays.copyOf(blobs, n * Vkzg.BYTES_PER_BLOB);
        Vkzg.computeCellKzgProofsBatch(slice, n); // warm up
        double best = Double.MAX_VALUE, total = 0;
        for (int r = 0; r < REPS; r++) {
            long start = System.nanoTime();
            Vkzg.computeCellKzgProofsBatch(slice, n);
            double ms = (System.nanoTime() - start) / 1_000_000.0;
            best = Math.min(best, ms);
            total += ms;
        }
        printRow(n, best, total / REPS);
    }

    private static void runRecoverCells(byte[] cells, byte[] present, int n) {
        final int cellsPerBlobBytes = Vkzg.CELL_PROOFS_PER_BLOB * Vkzg.BYTES_PER_CELL;
        byte[] cellsSlice = java.util.Arrays.copyOf(cells, n * cellsPerBlobBytes);
        byte[] presentSlice = java.util.Arrays.copyOf(present, n * Vkzg.CELL_PROOFS_PER_BLOB);
        Vkzg.recoverCellsBatch(cellsSlice, presentSlice, n); // warm up
        double best = Double.MAX_VALUE, total = 0;
        for (int r = 0; r < REPS; r++) {
            long start = System.nanoTime();
            Vkzg.recoverCellsBatch(cellsSlice, presentSlice, n);
            double ms = (System.nanoTime() - start) / 1_000_000.0;
            best = Math.min(best, ms);
            total += ms;
        }
        printRow(n, best, total / REPS);
    }

    private static void printRow(int n, double best, double avg) {
        System.out.printf(
                "  blobs: %3d   best: %9.2f ms   avg: %9.2f ms   per blob: %9.2f ms%n",
                n, best, avg, best / n);
    }

    // Deterministic pseudo-random canonical blob.
    private static void fillBlob(byte[] out, int offset, long seed) {
        long s = seed * 0x9E3779B97F4A7C15L + 1;
        for (int i = 0; i < Vkzg.FIELD_ELEMENTS_PER_BLOB; i++) {
            int feOff = offset + i * Vkzg.BYTES_PER_FIELD_ELEMENT;
            for (int j = 0; j < Vkzg.BYTES_PER_FIELD_ELEMENT; j++) {
                s = xorshift(s);
                out[feOff + j] = (byte) (s >>> 24);
            }
            // The modulus r's top byte is 0x73 (0b0111_0011); clearing just
            // its top two bits (rather than the whole byte) is enough to
            // guarantee every element is below r regardless of the
            // remaining bytes, while keeping 6 bits of real entropy.
            out[feOff] &= 0x3F;
        }
    }

    // Deterministic pseudo-random canonical full 128-cell extended array
    // (the same shape recoverCellsBatch expects: cell recovery never sees
    // an actual blob, only cells). Real, non-zero data matters here -- an
    // all-zero array would make every scalar multiply in the pipeline
    // degenerate and give misleadingly fast timings.
    private static void fillCells(byte[] out, int offset, long seed) {
        long s = seed * 0x9E3779B97F4A7C15L + 1;
        final int totalElements = Vkzg.CELL_PROOFS_PER_BLOB * Vkzg.FIELD_ELEMENTS_PER_CELL;
        for (int i = 0; i < totalElements; i++) {
            int feOff = offset + i * Vkzg.BYTES_PER_FIELD_ELEMENT;
            for (int j = 0; j < Vkzg.BYTES_PER_FIELD_ELEMENT; j++) {
                s = xorshift(s);
                out[feOff + j] = (byte) (s >>> 24);
            }
            out[feOff] &= 0x3F;
        }
    }

    private static long xorshift(long s) {
        s ^= s << 13;
        s ^= s >>> 7;
        s ^= s << 17;
        return s;
    }
}
